contracts/contracts/token/OUSD.sol

```diff
@@ -511,93 +491,181 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
                 _rebasingCredits,
                 _rebasingCreditsPerToken
             );
             return;
         }
 
-        _totalSupply = _newTotalSupply > MAX_SUPPLY
+        totalSupply = _newTotalSupply > MAX_SUPPLY
             ? MAX_SUPPLY
             : _newTotalSupply;
 
-        _rebasingCreditsPerToken = _rebasingCredits.divPrecisely(
-            _totalSupply.sub(nonRebasingSupply)
-        );
+        _rebasingCreditsPerToken =
+            (_rebasingCredits * 1e18) /
+            (totalSupply - nonRebasingSupply);
 
         require(_rebasingCreditsPerToken > 0, "Invalid change in supply");
 
-        _totalSupply = _rebasingCredits
-            .divPrecisely(_rebasingCreditsPerToken)
-            .add(nonRebasingSupply);
+        totalSupply =
+            ((_rebasingCredits * 1e18) / _rebasingCreditsPerToken) +
+            nonRebasingSupply;
 
         emit TotalSupplyUpdatedHighres(
-            _totalSupply,
+            totalSupply,
             _rebasingCredits,
             _rebasingCreditsPerToken
         );
     }
+
+    function delegateYield(address from, address to) external onlyGovernor {
+        require(from != address(0), "Zero from address not allowed");
+        require(to != address(0), "Zero to address not allowed");
+
+        require(from != to, "Cannot delegate to self");
+        require(
+            yieldFrom[to] == address(0) &&
+                yieldTo[to] == address(0) &&
+                yieldFrom[from] == address(0) &&
+                yieldTo[from] == address(0),
+            "Blocked by existing yield delegation"
+        );
+        RebaseOptions stateFrom = rebaseState[from];
+        RebaseOptions stateTo = rebaseState[to];
+
+        require(
+            stateFrom == RebaseOptions.NotSet ||
+                stateFrom == RebaseOptions.StdNonRebasing ||
+                stateFrom == RebaseOptions.StdRebasing,
+            "Invalid rebaseState from"
+        );
+
+        require(
+            stateTo == RebaseOptions.NotSet ||
+                stateTo == RebaseOptions.StdNonRebasing ||
+                stateTo == RebaseOptions.StdRebasing,
+            "Invalid rebaseState to"
+        );
+
+        if (alternativeCreditsPerToken[from] == 0) {
+            _rebaseOptOut(from);
+        }
+        if (alternativeCreditsPerToken[to] > 0) {
+            _rebaseOptIn(to);
+        }
+
+        uint256 fromBalance = balanceOf(from);
+        uint256 creditsFrom = _balanceToRebasingCredits(fromBalance);
+
+        // Set up the bidirectional links
+        yieldTo[from] = to;
+        yieldFrom[to] = from;
+        rebaseState[from] = RebaseOptions.YieldDelegationSource;
+        rebaseState[to] = RebaseOptions.YieldDelegationTarget;
+
+        // Local
+        _creditBalances[from] = fromBalance;
+        alternativeCreditsPerToken[from] = 1e18;
+        _creditBalances[to] += creditsFrom;
+
+        _adjustGlobals(creditsFrom.toInt256(), -fromBalance.toInt256());
+    }
```

Comment:
🟠 We need events at the ends of `delegateYield()`  and `undelegateYield()`.

Reply:
yes: https://github.com/OriginProtocol/origin-dollar/pull/2298/commits/504a4212b69f268ab0e05ef2b28c558b7e81e156